client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+
+In the case read returns None (value not in database), we still update our cache with the value entry None for the specified key at the read\_timestamp.  The None value may represent an update to the key.
+
+Note that the max\_age, and no\_cache parameters can be specified in at least three separate locations: (1) in the Cache constructor, (2) in the Cache read method, and (3) in the Transaction read method.  The cache read method determines the most restrictive set of parameters based on these three inputs and sends this most restrictive parameter set to the DB.  With respect to the read\_timestamp parameter, the DB server's response to every query includes the read\_timestamp it utilized to fulfill the request; we use the read\_timestamp the server returns to us to update the cache.
+
+Here is the method for determining the most restrictive parameter set:
+
+* **read\_timestamp**: The most restritive read\_timestamp is the minimum of the given read\_timestamp values (i.e., min(cache, transaction) read\_timestamps)
+* **max\_age**: The most restrictive max\_age is the minimum of all given max\_age values
+* **no\_cache**: The most restrictive no\_cache value is the disjunction (boolean "or") of all given no\_cache values.  (If no\_cache == True, the cache must verify with the server by checking the value timestamp it has the most up-to-date value.)
+
+### write
+    write(
+        table: string, 
+        op_List: list): value (JSON) throws StaleException  
+
+The write method sends the op\_List to the DB server as a batch write.  
+
+If the server accepts the batch write, the write method also updates the cache to reflect the newly written values.
+
+* **Question (2)**: What is the expected format of the op\_List?  (What are the possible types of its elements?)
+
+* **Question (3)**: Should the write method update the cache upon a successful write (with the newly written values)?
+
+Otherwise, the write method throws a stale exception, including the value timestamp in the DB that caused the write reject if it is available.
+
+## HTTP Message Formatting
+First, we specify the default and custom headers our HTTP messages utilize.  Then we describe the expected forms of requests/responses between the cache and TreodeDB server.
+
+### Use of HTTP Headers
+
+In our use of HTTP headers, the term *age* refers to time since the user-specified read\_timestamp.  Specifically, every TreodeDB entry has both a *value timestamp (V_t)* and *read timestamp (R_t)*.  We maintain as an invariant that the given key contained the entry's value between [V\_t, R\_t].
+
+We use the following HTTP headers:
+
+* **max-age**: The max number of seconds between our given read\_timestamp and the read\_timestamp of the entry, i.e., (input read\_timestamp) - (entry read\_timestamp) <= max-age.  In other words, we are willing to read a value that could be stale by at most max-age seconds. 
```

Comment:
Let's be careful how we talk about this, so that somebody confused because they are new to this stuff isn't further confused by our casual talk. The HTTP _header_ is `Cache-Control`. `max-age` and `no-cache` are _directives_ that can be used with that header. For example:

    GET /table/key HTTP/1.1
    Cache-Control: max-age=3